### PR TITLE
SSD Bot

### PR DESCRIPTION
**Pull request by Kyep**
Adds a SSD retrieval bot.

When spawned in by an admin, it will sit in cryodorms, and, every few minutes, check for SSD people. If it finds them, and they meet a long list of conditions, it goes to them, picks them up, brings them back, and moves them to cryo.

Why is this a good thing? Many reasons:
A) Antags have a better time, because their SSD targets get replaced faster.
B) Admins have an easier time, because it should drastically reduce the # of ahelps we get about interactions with SSDs (because there willb e less SSD characters in the round).
C) Everyone has a better time, because with SSD people put in cryo more easily, more job slots are open and everyone has a better chance of their favorite job being available when joining late.

Currently, the bot will only seek out and retrieve characters who are:
- On the station zlevel
- SSD
- Alive
- Humanoid crew
- Not anchored
- Not stunned
- Not handcuffed
- Not in cryodorms already
- Not in space
- Not buckled
- Not pulled, dragged or grabbed
- In an area which the bot can access (ie: not virology, not xenobio)
- In an area where people have no incentive to go SSD to get out (ie: not brig cells or perma).
- Not a head of staff
- Is not only SSD, but has been so for several minutes.

Once all these conditions are met, the bot physically travels to them, picks them up, carries them back to the cryodorms, and if a pod is available, puts them into cryo.

Again, this both should make SSD handling much better, with benefits for everyone. It should be low risk, as this PR does not map the bot in. It simply allows admins to spawn it, and observe how it works. If it turns out to be popular, it can be added to the default map at a later date.

Thread: http://nanotrasen.se/phpBB3/viewtopic.php?f=12&t=9220

### CODE REVIEW

**Comment by Aurorablade** [comment]
I am still opposed to this and will remain opposed to this as we have: the HoP/Captain to oversee new job slots opening, You really need admins invovled for SSD handleing issues short of putting them into cryo.

If anything the better solution would be for admins if they HAD TO is making a command to put a specific character into cryo and a command to empty a cryopod to force cycle it if needed. you could also make a system to scale job slots based on population. but again, Captain and Hop.

File: code/modules/mob/living/simple_animal/bot/ssdbot.dm [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+	var/list/areas_to_ignore = list(/area/medical/virology, /area/toxins/xenobiology, /area/security/prison, /area/security/permabrig)
+	var/pickup_delay = 60 // 60 cycles, 120 seconds
+	var/list/previous_targets = list()
+	var/mob/targetssd = null
+
+/mob/living/simple_animal/bot/mulebot/ssdbot/New()
+	..()
+	name = initial(name)
+	home_turf = get_turf(src)
+	var/datum/job/captain/J = new/datum/job/captain
```
Comment (tigercat2000) [inline code/modules/mob/living/simple_animal/bot/ssdbot.dm]: What the actual fuck?
This should just be `access_card.access = get_all_accesses()`
You should never directly create a job datum.

File: code/modules/mob/living/simple_animal/bot/ssdbot.dm [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+	ignored_ssds += H
+	target = get_turf(H)
+	targetssd = H
+	speak("Getting directions to [H] at <b>[get_area(H)]</b>.")
+	var/area/dest_area = get_area(target)
+	destination = format_text(dest_area.name)
+	start()
+
+/mob/living/simple_animal/bot/mulebot/ssdbot/proc/bot_can_pick_up(var/mob/living/carbon/human/H)
+	if(H.z == z && isLivingSSD(H) && !H.anchored && !H.stunned && !H.handcuffed && H.loc != home_turf && !istype(get_turf(H), /turf/space) && !H.buckled && !H.pulledby && !H.grabbed_by)
```
Comment (tigercat2000) [inline code/modules/mob/living/simple_animal/bot/ssdbot.dm]: Seperate these conditionals into multiple lines, please.

File: code/modules/mob/living/simple_animal/bot/ssdbot.dm [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+	if(!reached_target)
+		if(soundeffects)
+			audible_message("[src] makes a chiming sound!", "<span class='emote'>You hear a chime.</span>")
+			playsound(loc, 'sound/machines/chime.ogg', 50, 0)
+		reached_target = 1
+
+		if(load)
+			speak("Unloading [load].")
+			var/mob/living/carbon/human/N = passenger
+			unload(2) // south
```
Comment (tigercat2000) [inline code/modules/mob/living/simple_animal/bot/ssdbot.dm]: ... use the directional defines. `unload(SOUTH)`

File: code/modules/mob/living/simple_animal/bot/ssdbot.dm [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+	to_chat(user, "<span class='warning'>[src] has no visible maintenance hatch.</span>")
+
+/mob/living/simple_animal/bot/mulebot/Move(turf/simulated/next)
+	..()
+	if(buckled_mob)
+		buckled_mob.dir = dir
+
+/mob/living/simple_animal/bot/mulebot/ssdbot/RunOver(mob/living/carbon/human/H)
+	H.Weaken(5)
+	return // this version doesn't hurt people even if it runs into them
```
Comment (tigercat2000) [inline code/modules/mob/living/simple_animal/bot/ssdbot.dm]: No trailing returns.

**Comment by Fox-McCloud** [comment]
I too am still opposed to this; as pointed out, it's going to create more problems than it actually resolves. Furthermore, I don't exactly trust this behavior to a bot, to say the least--they're prone to breaking are laggy, and cause performance issues. If having a ton of SSD people are really *that* concerning to administration, then it'd be far better to just have a cleanup controller than poofs people away who have been SSD for a long time, or opens up further slots after someone has been SSD for a long while.

**Comment by Lady-Luck** [comment]
Wouldn't it be better to just have some sort of passive SSD detection where if someone has been SSD for more then 15 minutes, that job's slot opens and traitors get a new target?

**Comment by Allfd** [comment]
It could be possible that the bot only responds if someone's medical record has them as SSD,  that way it is up to the captain/hop/medical

**Comment by Kyep** [comment]
@tigercat2000 updated

**Comment by Vivalas** [comment]
As long as no rules get made against destroying the bot then sure I guess. Seems like all those checks would be laggy but it's only every 5 minutes or so, won't be that big of a problem.

**Comment by Kyep** [comment]
Not popular enough to get merged. Closing.

**Pull request closed without merging**